README.md
# battlesim: Modelling and animating simulated battles between units in Python.

![pypi](https://img.shields.io/pypi/v/battlesim)
![last commit](https://img.shields.io/github/last-commit/gregparkes/BattleSimulator)
![repo size](https://img.shields.io/github/repo-size/gregparkes/BattleSimulator)

The aim of this side project is to become familiar with Python classes and with 
primitive forms of animation and simulating environments. We map units onto a
2D plane and run simple simulations that involve them moving towards an enemy 
unit and attacking it. Rounds finish when one team has completely wiped out
the other side, or we have reached the maximum number of timesteps.

Users familiar with [Totally Accurate Battle Simulator](https://steamcommunity.com/app/508440) will hopefully 
love this package as a lot of the basic ideas are derived from this.

![Image not found](simulations/main.gif)

The code for the primary engine is found in `battlesim/`, and implementations/
examples are found in the Jupyter notebooks. Animations should display
properly in these notebooks.

## Installation

`battlesim` requires the following [dependencies](environment.yml):

* python (>=3.7)
* numpy (>=1.11.0)
* pandas (>=0.25.1)
* matplotlib (>=3.1.1)
* numba (>=0.45)

With the following for exporting the animation as a gif:

* ffmpeg (>=4.2)

The following packages are not required but significantly improve the usage of this package. If you are unfamiliar with the Jupyter project see [here](https://jupyter.org/):

* jupyter (1.0.0)

### From PyPI

If you have working versions of the dependencies, similarly [install using pip](https://pypi.org/project/battlesim/) (version 0.3.5):

```bash
pip install battlesim
```

We recommend updating the dependencies yourself using conda 
rather than through pip because conda manages the dependencies
better, but pip will do it for you. See the `environment.yml` 
file for dependencies.

### From Cloning the GitHub Repository

Alternatively if you are cloning this [GitHub repository](https://github.com/gregparkes/BattleSimulator), use:

```bash
git clone https://github.com/gregparkes/BattleSimulator
conda env create -f environment.yml
conda activate bsm
```

Now within the `bsm` environment run your Jupyter notebook:

```bash
jupyter notebook
```

### Running Tests

You will need the following for testing (soft requirement):

* PyTest (5.1.2)

Then perform the following within a console:

```bash
cd tests/
pytest -v
```

## How to use: The Basics

Firstly, check the requirements for using this simulator, of which most come with the [Anaconda distribution](https://www.anaconda.com/). In addition you will need the **ffmpeg** video conversion package to generate the simulations as animations.

Secondly, you will need to import the package as:

```python
import battlesim as bsm
```

We recommend using `bsm` as a shorthand to reduce the amount of writing out you have to do. If you're using Jupyter notebook we also recommend:

```python
import matplotlib.pyplot as plt
plt.rcParams["animation.html"] = "html5"
```

The second line is important when you come to plotting the animations, as there are a number of issues with using it. All of the heavy lifting comes in the `bsm.Battle` object that provides a neat interface for all of the operations you would like to conduct:

```python
bat = bsm.Battle("datasets/starwars-clonewars.csv")
```

You can see that we have specified a 'dataset' from which all of
the unit roster can be drawn from; for specifics of how this file should
be oriented, see the documentation. We then need to specify 
units to create to form an army. For example, in this Star Wars 
example, we could specify a play-off between Clone troopers 
and B1 battledroids.

This is achieved using a meta-information object called a `Composite`, which holds
a group of units of a given type.

```python
armies = [
    bsm.Composite("B1 battledroid", 70),
    bsm.Composite("Clone Trooper", 50)
]
bat.create_army(armies)
```

which internally creates an efficient `numpy` matrix, 
ready to perform the simulation. This is stored in the `battle.M_` 
object, a heterogenous `ndarray` element. By default, each Composite spawns
on top of each other using a gaussian distribution at (0, 0). When initialising
the Composite we can specify a new sampling using the `Sampling` class or
override directly:

```python
bat.composition_[1].pos = bsm.Sampling("normal", 10., 2.)
```

And now to simulate (note that the first time this executes will be painfully slow
as JIT compiles a lot of code):

```python
F = bat.simulate()
```

By default, the simulation function will make a record of
important parameters at each step and then return these 
parameters as a heterogenous `np.ndarray` at the end 
in *long form* (with a cached element called `sim_`). 
In addition, because you want to see what's going on - we can 
animate the frames using this convenience method within
the Battle object:

```python
bat.sim_jupyter()
```

![Image not found](simulations/sim2.gif)

Here `sim_jupyter` treats each unit object as a quiver 
arrow in 2-d space (position and direction facing it's enemy).
The targets should move towards each other and attempt to kill
each other. Dead units are represented as crosses **'x'** on the map. 

![Image not found](images/quiver1.svg)

The rest is for you to explore, tweak and enjoy watching arrows
move towards each other and kill each other. We have 
extensive [examples](https://github.com/gregparkes/BattleSimulator/tree/master/examples) to look at within this repository.

## One step further: Repeated runs

If you're interested in seeing how each team fare over
multiple runs (to eliminate random biases), then `bsm.Battle` 
objects once defined, contain a `simulate_k()` method, 
where `k` specifies the number of runs you wish to complete.
Unlike `simulate()` by itself, it does not 
return a `np.ndarray` of frames, but rather the 
number of units from each team left standing at each iteration.

```python
>>> runs = battle.simulate_k(k=40)
```

This is the beginning of creating an interface similar 
to Machine Learning, whereby the outcome can be a 
classification (team) or regression 
(number of units surviving) target, and the unit 
compositions, aspects of the engine etc., can be inputs.

## New in v0.3.6

There are a number of exciting changes in this [current update](CHANGELOG.md), including:

- Introduction of **Terrains**. This is a major expansion giving 3D pseudodepth to animated battles. Depth now influences movement speed of units, with terrain penalties applied (up to 50%) on higher hills. They also increase range for units on hills and increase damage when firing downhill on an enemy unit.
- Introduction of *armor*. Armor acts as another health buffer to protect units from harm.

## Teaching series

As well as a fully-fledged package simulator, you can find
[teaching material](https://github.com/gregparkes/BattleSimulator/tree/master/teaching) in Jupyter 
notebook form within the `teaching/` subfolder, that takes 
users through the development process of this package, compares
and contrasts Object-Oriented (OO) implementations to numpy-esque 
implementations, their performance, plotting, animations and more. 
We hope you find this material interesting and will aid as you use 
the package and possibly develop packages of your own in the future.

Material covered so far:

1. Basics, including importing the dataset, the `Unit` class, basic simulation
2. Improving the `Unit` class and simulation early-stopping for performance.
3. Plotting simulations and performance-driven development

This is still in active development retracing the steps of the project. All legacy functions associated with this can be found in the `battlesim/legacy.py` document.

## Future plans

* Include AI-based behavior that makes use of height (to occupy hills)
* Develop 'defensive' AI.

***

Ensure that any use of this material is appropriately referenced 
and in compliance with the [license](LICENSE.txt).

[![Run on Repl.it](https://repl.it/badge/github/gregparkes/BattleSimulator)](https://repl.it/github/gregparkes/BattleSimulator)

CHANGELOG.md
# ChangeLog

All notable changes to this project will be documented in this file.

## [0.3.7] - 2020-10-23
### Added
- `Composite` class to create armies with. This holds all the meta information for a unit group.
- `Sampling` class. Lightweight replacement for `Distribution`.
- Optional `tqdm` package support with simulate_k method.

### Changed
- Significant internal cleaning of `Battle` object.
- Optimisations to `simulate_fast` function have lead to
around 20x speed up using `numba` more efficiently.
- Integration of numpy array with heterogenous aligned data type with numba continued.
- Unit tiling significantly improved using lerp over distance minimizations.

### Removed
- `Distribution` class. This was too verbose and has been replaced with `Sampling`.


## [0.3.6] - 2020-02-04
### Added
- Terrains can now be generated using *perlin noise*. This is not only significantly faster for larger maps but generates more attractive background maps.

### Changed
- Large-scale internal adjustments, code cleaning, new folders and files
- Latest versions of `numba` allow for heterogeneous matrix dtypes, allowing for code refactoring for jit sections. Significantly cleaner function arguments.
- adjustments as to how `Terrain` is generated.

## [0.3.5] - 2019-10-18
### Added
- `Terrain`. This is a major expansion giving 3D pseudodepth to animated battles. Depth now influences:
	- Decreased movement speed on hills
	- Increased range of units on hills
	- Increased damage output of units targeting downhill enemies
- Boundary checking for units
- `pip` support
- `CHANGELOG.md` file
- First release on GitHub 0.3.5
- `jitcode` for streamlined `numba` use
- Global `target` functions for initialization
- Armor calculations for units
- `example4.ipynb` to reflect changes

### Changed
- `__init__` file dependencies
- Removed file requirement for `Battle()`
- `Distribution` constructor now accepts copy-constructor of another Distribution
- Made `Battle` attributes properties to protect them

### Removed
- Alternative constructor for `Distribution`


## [0.3.4] - 2019-10-11
### Added
- Integrated terrain background
- Terrain weighting in unit movement calculations

### Changed
- Streamlined `Terrain`, `Distribution` into `Battle` object

### Removed
- `test_move` test suite


## [0.3.2] - 2019-10-07
### Added
- Teaching material
- General *decision-based AI* with `aggressive` and `hit_and_run` options
- `Terrain` definition
- LICENSE.

### Changed
- Size-dependency on quiver-plots to units.
- Streamlined legend
- Memory usage from `M_` object

### Removed
- `set_position` in `Battle`

***

Ensure that any use of this material is appropriately referenced and in compliance with the license.

battlesim/_ai.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  8 11:07:18 2019

@author: gparkes
"""
from numba import njit

from . import _move
from . import _hit
from . import _damage


def get_function_names():
    """Returns the function names."""
    return "aggressive", "hit_and_run"


def get_functions():
    """Returns the functions themselves."""
    return aggressive, hit_and_run


def get_map_functions():
    """Maps the function names to the functions."""
    return dict(zip(get_function_names(), get_functions()))


__all__ = get_function_names()

"""
Parameters:
    pos: (np.ndarray, (n, 2), float all) positions of every unit
    speed: (np.ndarray, (n,) float all) base speed of every unit
    mrange: (np.ndarray, (n,) float all) base range of every unit
    acc: (np.ndarray, (n,) float all) accuracy of every unit
    dodge: (np.ndarray, (n,) float all) dodge of every unit
    target: (np.ndarray, (n,) int all) id of every target
    dmg: (np.ndarray, (n,) float all) base damage of every unit
    hp: (np.ndarray, (n,) float all) current HP of every unit
    armor: (np.ndarray, (n,) float all) current armor of enemy unit
    luck: (np.ndarray, (n,2) float all) luck roll of every unit
    distances: (np.ndarray, (n,) float all) unit distance to target
    dd: (np.ndarray, (n,) float all) directional derivative to target
    team: (np.ndarray, (n, ) int all) team ID of every unit
    target_f: (function) target function.
    enemies: (np.ndarray, (e,) int all) IDs of every enemy unit
    allies: (np.ndarray, (a,) int all) IDs of every allied unit
    Z: (np.ndarray, (res, res) float all) the terrain of the map
    Z_xi: (np.ndarray, (n,) int all) the x-index of Z-terrain tile
    Z_yi: (np.ndarray, (n,) int all) the y-index of Z-terrain tile
    i : (int) the current unit's index
"""


@njit
def _height_modification(h):
    """
    height modifier follows:
        h_{mod} = 1 + (h^2 / 3), h in [0, 1]
    """
    return (h ** 2 / 3.) + 1.


@njit
def _select_enemy(M, f_t, enemies, allies, i):
    if M['hp'][M['target'][i]] <= 0:
        if enemies.shape[0] > 0:
            t = f_t(M['x'], M['y'], M['hp'], enemies, allies, i)
            if t != -1:
                M['target'][i] = t
                return True
            else:
                return False
        else:
            return False
    else:
        return True


@njit
def _rng_terr(mrange, Z, Z_xi, Z_yi, i):
    """ The range of the unit, modified by the terrain measurement. Returns float. """
    return mrange[i] * _height_modification(Z[Z_xi[i], Z_yi[i]])


""" -------------------------- ACCESSED FUNCTIONS -----------------------------------------"""


# pos, speed, mrange, acc, dodge, targets,
#               dmg, hp, armor, luck, distances, dd, team, target_f,
#               enemies, allies, Z, Z_xi, Z_yi, i


@njit
def aggressive(M, luck, dists, dx, dy, target_f,
               enemies, allies, Z, Z_xi, Z_yi, i):
    """
    This basic AI looks whether the current unit i is in range of it's target'
    and if it isn't, moves towards it until it is, then attacks.
    """

    """# use ai_map to dictionary-map the group number to the appropriate AI function"""
    if _select_enemy(M, target_f, enemies, allies, i):
        # if not in range, move towards target, or hit a chance (5%) and move forward anyway.
        if dists[i] > _rng_terr(M['range'], Z, Z_xi, Z_yi, i) or luck[i] < 0.05:
            """# move unit towards attacking enemy."""
            _move.to_enemy(M, dx, dy, dists, Z, Z_xi, Z_yi, i)
            return True
        else:
            """# calculate the chance of hitting the opponent"""
            h_chance = _hit.basic_chance(M, dists, i)
            """if hit chance overcomes round luck.. deal damage to HP."""
            if h_chance > luck[i]:
                _damage.basic(M, Z, Z_xi, Z_yi, i)
            return True
    else:
        return False


@njit
def hit_and_run(M, luck, dists, dx, dy, target_f,
                enemies, allies, Z, Z_xi, Z_yi, i):
    """
    This AI option sees if its range/movement is greater than it's enemy, and if it is, it
    performs hit-and-run on it's opponent.
    """
    # assign target enemy
    if _select_enemy(M, target_f, enemies, allies, i):

        # cache quick stats
        t_i = M["target"][i]
        range_i = _rng_terr(M["range"], Z, Z_xi, Z_yi, i)
        range_j = _rng_terr(M["range"], Z, Z_xi, Z_yi, t_i)

        if (M["speed"][i] > M["speed"][t_i]) and (range_i > range_j):

            # if we're out of range, move towards
            if dists[i] > range_i:
                """# move towards unit."""
                _move.to_enemy(M, dx, dy, dists, Z, Z_xi, Z_yi, i)
                return True
            # else if the enemy is in range, back off
            elif dists[i] < range_j:
                """# move directly away from unit."""
                _move.from_enemy(M, dx, dy, dists, Z, Z_xi, Z_yi, i)
                return True
            else:
                """# so we're in range, the enemy is not, attack."""
                h_chance = _hit.basic_chance(M, dists, i)

                if h_chance > luck[i]:
                    _damage.basic(M, Z, Z_xi, Z_yi, i)
                return True
        else:
            # otherwise just perform an 'aggressive' model.
            return aggressive(M, luck, dists, dx, dy, target_f,
                              enemies, allies, Z, Z_xi, Z_yi, i)
    else:
        return False


def defensive(pos, speed, mrange, acc, dodge, targets,
              dmg, hp, armor, luck, distances, dd, team, target_f,
              enemies, allies, Z, Z_xi, Z_yi, i):
    """
    This AI option attempts to find a nearby high hill and sit on it, waiting for
    a nearby enemy.
    """
    # use target index to see if we can reach the hill in-time.
    return NotImplemented
